src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -344,133 +344,312 @@ public class SqlMutationEngine : IMutationEngine
+                }
+            }
+
+            // READ action for the given role can have include and exclude fields configured. Populating UpdateReturnFields
+            // ensures that the select query retrieves only those fields that are allowed for the given role.
+            findRequestContext.UpdateReturnFields(_authorizationResolver.GetAllowedExposedColumns(context.EntityName, roleName, EntityActionOperation.Read));
+
+            return findRequestContext;
+        }
+
+        /// <summary>
+        /// Constructs and returns a HTTP 201 Created response.
+        /// The response is constructed using results of the upsert database operation when database policy is not defined for the read permission.
+        /// If database policy is defined, the results of the subsequent select statement is used for constructing the response.
+        /// </summary>
+        /// <param name="resultRow">Reuslt of the upsert database operation</param>
+        /// <param name="jsonDocument">Result of the select database operation</param>
+        /// <param name="isReadPermissionConfiguredForRole">Indicates whether read permissions is configured for the role</param>
+        /// <param name="isDatabasePolicyDefinedForReadAction">Indicates whether database policy is configured for read action</param>
+        private static CreatedResult ConstructCreatedResultResponse(Dictionary<string, object?> resultRow, JsonDocument? jsonDocument, string primaryKeyRoute, bool isReadPermissionConfiguredForRole, bool isDatabasePolicyDefinedForReadAction)
+        {
+            // When the database policy is defined for the read action, a select query in another roundtrip to the database will be executed to fetch the results.
+            // So, the response of that database query is used to construct the final response to be returned.
+            if (isDatabasePolicyDefinedForReadAction)
+            {
+                return (jsonDocument is not null) ? new CreatedResult(location: primaryKeyRoute, OkMutationResponse(jsonDocument.RootElement.Clone()).Value)
+                                                  : new CreatedResult(location: primaryKeyRoute, OkMutationResponse(JsonDocument.Parse("[]").RootElement.Clone()).Value);
+            }
+
+            // When no database policy is defined for the read action, the results from the upsert database operation is
+            // used to construct the final response.
+            // When no read permission is configured for the role, or all the fields are excluded
+            // an empty response is returned.
+            return (isReadPermissionConfiguredForRole && resultRow.Count > 0) ? new CreatedResult(location: primaryKeyRoute, OkMutationResponse(resultRow).Value)
+                                                     : new CreatedResult(location: primaryKeyRoute, OkMutationResponse(JsonDocument.Parse("[]").RootElement.Clone()).Value);
```

Comment:
While returning an empty response, if we provide a valid primaryKeyRoute in the location header even if read permission is not configured, we might provide more information unnecessarily

Reply:
True. 

My thought process behind returning a valid location header is that if the client application wants to perform more operations for which it has permissions configured, it would be able to do so. 

For example, consider flows such as below:
1) Creating a new record 
2) After some business logic execution, one or more fields have to be updated.
(or)
3) Performing a subsequent READ with a different role which has the read permissions.

Without the location header in the response, we would not be able to enable these scenarios.

Reply:
For PUT and PATCH API requests, the users are aware of the Pks as it is required to be passed in the request URL.

In case of tables with auto-gen PKs, PUT or PATCH will not result in an insert but error out. Seeing that Location Header does not provide users with any additional information, it is set as an empty string always.

For POST API requests, the primary key route calculated will be an empty string only when the read action for the role does not have access to one or more PKs. So, in this case too, the Location Header is set to an empty string.

Reply:
After offline conversation, decided to give empty location header for POST response if read is not configured. Its the developer's responsibility to configure appropriate roles.